pkg/cmd/server/kubernetes/master/master_config.go

```diff
@@ -163,57 +170,30 @@ func BuildKubeAPIserverOptions(masterConfig configapi.MasterConfig) (*kapiserver
 }
 
 // BuildStorageFactory builds a storage factory based on server.Etcd.StorageConfig with overrides from masterConfig.
-// This storage factory is ONLY USED FOR kubernetes registries right now. Compare pkg/util/restoptions/configgetter.go.
-func BuildStorageFactory(masterConfig configapi.MasterConfig, server *kapiserveroptions.ServerRunOptions) (apiserverstorage.StorageFactory, error) {
+// This storage factory is used for kubernetes and origin registries. Compare pkg/util/restoptions/configgetter.go.
+func BuildStorageFactory(masterConfig configapi.MasterConfig, server *kapiserveroptions.ServerRunOptions, enforcedStorageVersions []schema.GroupVersionResource) (*apiserverstorage.DefaultStorageFactory, error) {
 	resourceEncodingConfig := apiserverstorage.NewDefaultResourceEncodingConfig(kapi.Registry)
-	resourceEncodingConfig.SetVersionEncoding(
-		kapi.GroupName,
-		schema.GroupVersion{Group: kapi.GroupName, Version: masterConfig.EtcdStorageConfig.KubernetesStorageVersion},
-		kapi.SchemeGroupVersion,
-	)
-
-	resourceEncodingConfig.SetVersionEncoding(
-		extensions.GroupName,
-		schema.GroupVersion{Group: extensions.GroupName, Version: "v1beta1"},
-		extensions.SchemeGroupVersion,
-	)
-
-	resourceEncodingConfig.SetVersionEncoding(
-		batch.GroupName,
-		schema.GroupVersion{Group: batch.GroupName, Version: "v1"},
-		batch.SchemeGroupVersion,
-	)
-
-	resourceEncodingConfig.SetVersionEncoding(
-		autoscaling.GroupName,
-		schema.GroupVersion{Group: autoscaling.GroupName, Version: "v1"},
-		autoscaling.SchemeGroupVersion,
-	)
 
 	storageGroupsToEncodingVersion, err := server.StorageSerialization.StorageGroupsToEncodingVersion()
 	if err != nil {
 		return nil, err
 	}
+	for group, storageEncodingVersion := range storageGroupsToEncodingVersion {
+		resourceEncodingConfig.SetVersionEncoding(group, storageEncodingVersion, schema.GroupVersion{Group: group, Version: runtime.APIVersionInternal})
+	}
+	resourceEncodingConfig.SetResourceEncoding(batch.Resource("cronjobs"), batchv2alpha1.SchemeGroupVersion, batch.SchemeGroupVersion)
 
-	// use the stock storage config based on args, but override bits from our config where appropriate
-	etcdConfig := server.Etcd.StorageConfig
-	etcdConfig.Prefix = masterConfig.EtcdStorageConfig.KubernetesStoragePrefix
-	etcdConfig.ServerList = masterConfig.EtcdClientInfo.URLs
-	etcdConfig.KeyFile = masterConfig.EtcdClientInfo.ClientCert.KeyFile
-	etcdConfig.CertFile = masterConfig.EtcdClientInfo.ClientCert.CertFile
-	etcdConfig.CAFile = masterConfig.EtcdClientInfo.CA
+	// use legacy group name "" for all resources that existed when apigroups were introduced
+	for _, gvr := range enforcedStorageVersions {
```

Comment:
we pass `nil` for `enforcedStorageVersions` when called from `buildKubeApiserverConfig`... is that explicitly for origin resources?